### ISSUE TITLE
Documentation with `when` branch is missing field types

### ISSUE THREAD

**Issue by mratsim** [nim-lang/Nim#9899]
Test case: The current `options` module documentation is missing type for fields: 

https://nim-lang.org/docs/options.html#7

```Nim
Option[T] = object
  when T is SomePointer:
    val
  else:
      val
      has
```

While this is what is declared:

https://github.com/nim-lang/Nim/blob/72e15ff739cc73fbf6e3090756d3f9cb3d5af2fa/lib/pure/options.nim#L75-L82

**Comment by nc-x** [comment]
This is a general issue not related to docgen:
```
proc x(n: int): string = discard

x:
  type
    Option*[T] = object
      ## An optional type that stores its value and state separately in a boolean.
      when T is void:
        val: T
      else:
        val: T
        has: bool
```
`nim c t.nim` gives -
```
t.nim(3, 1) Error: type mismatch: got <void>
but expected one of:
proc x(n: int): string

expression: x:
  type
    Option[T] = object
      when T is void:            ## An optional type that stores its value and state separately in a boolean.
        val
      else:
          val
          has
```
 -----
  
This is because `semRecordNodeAux` for `nkRecWhen` for `generics` evaluates stuff early.

A simple fix is to remove this line https://github.com/nim-lang/Nim/blob/9c3e23e0751881b16fab813fd3c2edc1b98b7f68/compiler/semtypes.nim#L676
but then obviously it creates other issues.
 
  -----
   
Another fix would be to check for this case here -
https://github.com/nim-lang/Nim/blob/9c3e23e0751881b16fab813fd3c2edc1b98b7f68/compiler/docgen.nim#L354-L356
and use `PSym_variable.typ` to get what we want. (And similarly for renderer or whatever the compiler uses to print in errors).

There may be other ways to fix it, IDK.

cc @Araq

**Comment by metagn** [comment]
Fixed by #21017
